File: src/core/engine.ts
import { BlockPattern, CleanedResult, CleanupRuleSet, RawInput, SourceType } from './types';
import { detectSourceType } from './detector';
import { GenericRuleSet } from './rules/generic';
import { GitHubRuleSet } from './rules/github';
import { DocsRuleSet } from './rules/docs';
import { ArticleRuleSet } from './rules/article';
import { ChatRuleSet } from './rules/chat';

export function getRuleSetForType(type: SourceType): CleanupRuleSet {
  if (type === 'github_pr' || type === 'github_issue') {
    return GitHubRuleSet;
  }
  if (type === 'docs') {
    return DocsRuleSet;
  }
  if (type === 'article') {
    return ArticleRuleSet;
  }
  if (type === 'chat') {
    return ChatRuleSet;
  }
  return GenericRuleSet;
}

export function normalizeText(text: string): string[] {
  // Normalize line endings to \n, trim outer whitespace, split into lines.
  // Strip U+FFFC (Object Replacement Character) — web pages replace images/icons
  // with this character when copied as plain text; it is always noise.
  // Normalize U+00A0 (No-Break Space) to a regular space — avoids mismatches
  // against rule strings that use ordinary spaces.
  return text.replace(/\r\n/g, '\n').trim().split('\n')
    .map(line => line.replace(/\uFFFC/g, '').replace(/\u00A0/g, ' '));
}

export function isPreserved(line: string, ruleSet: CleanupRuleSet): boolean {
  return ruleSet.preserveRegexes.some(regex => regex.test(line));
}

export function trimPrefix(lines: string[], ruleSet: CleanupRuleSet): string[] {
  let unrecognizedCount = 0;
  let lastJunkIndex = -1;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (line === '') continue;

    let isJunk = false;
    if (ruleSet.prefixExactLines.includes(line)) {
      isJunk = true;
    } else if (ruleSet.prefixRegexes.some(regex => regex.test(line))) {
      isJunk = true;
    }

    if (isJunk) {
      lastJunkIndex = i;
      unrecognizedCount = 0;
    } else {
      if (isPreserved(line, ruleSet)) {
        break;
      }
      unrecognizedCount++;
      if (unrecognizedCount > 5) {
        break;
      }
    }
  }
  
  if (lastJunkIndex >= 0) {
    return lines.slice(lastJunkIndex + 1);
  }
  return lines;
}

export function trimSuffix(lines: string[], ruleSet: CleanupRuleSet): string[] {
  let unrecognizedCount = 0;
  let firstJunkIndex = lines.length;

  for (let i = lines.length - 1; i >= 0; i--) {
    const line = lines[i].trim();
    if (line === '') continue;

    let isJunk = false;
    if (ruleSet.suffixExactLines.includes(line)) {
      isJunk = true;
    } else if (ruleSet.suffixRegexes.some(regex => regex.test(line))) {
      isJunk = true;
    }

    if (isJunk) {
      firstJunkIndex = i;
      unrecognizedCount = 0;
    } else {
      if (isPreserved(line, ruleSet)) {
        break;
      }
      unrecognizedCount++;
      if (unrecognizedCount > 5) {
        break;
      }
    }
  }
  
  if (firstJunkIndex < lines.length) {
    return lines.slice(0, firstJunkIndex);
  }
  return lines;
}

/**
 * Pre-compute a set of line indices that must survive all cleanup steps.
 *
 * This is the "protect specific blocks" stage that runs **before** any removal
 * step.  It identifies two categories of protected content:
 *
 *  1. Fenced code blocks (``` … ```) — always protected regardless of rule set.
 *  2. Blocks matching `ruleSet.preserveBlockPatterns` — rule-defined content
 *     that gives the downstream LLM valuable context (e.g. diff hunks, review
 *     explanation paragraphs).
 *
 * Every line whose index is in the returned set is immune from all subsequent
 * removal steps (`removeBlocks`, `cleanMiddle`).
 */
export function computeProtectedLines(lines: string[], ruleSet: CleanupRuleSet): Set<number> {
  const protectedSet = new Set<number>();

  // Step 1 — always protect fenced code blocks
  let inCodeBlock = false;
  for (let i = 0; i < lines.length; i++) {
    const trimmed = lines[i].trim();
    if (trimmed.startsWith('```')) {
      protectedSet.add(i); // fence line itself is always protected
      inCodeBlock = !inCodeBlock;
    } else if (inCodeBlock) {
      protectedSet.add(i);
    }
  }

  // Step 2 — rule-defined preserve block patterns
  const patterns = ruleSet.preserveBlockPatterns;
  if (!patterns || patterns.length === 0) return protectedSet;

  let i = 0;
  while (i < lines.length) {
    if (protectedSet.has(i)) { i++; continue; }
    const trimmed = lines[i].trim();
    let advanced = false;

    for (const bp of patterns) {
      if (bp.start.test(trimmed)) {
        const maxLen = bp.maxLines ?? 80;
        protectedSet.add(i); // protect the start line itself

        if (bp.end) {
          let j = i + 1;
          while (j < lines.length && (j - i) < maxLen) {
            protectedSet.add(j);
            if (bp.end.test(lines[j].trim())) {
              i = j + 1;
              advanced = true;
              break;
            }
            j++;
          }
          if (!advanced) {
            i = j; // end not found within maxLines — advance past
            advanced = true;
          }
        } else {
          // No end pattern: extend to next blank line (or maxConsecutiveBlankLines)
          const maxBlanks = bp.maxConsecutiveBlankLines ?? 1;
          let j = i + 1;
          let consecutiveBlanks = 0;
          while (j < lines.length && (j - i) < maxLen) {
            if (lines[j].trim() === '') {
              consecutiveBlanks++;
              if (consecutiveBlanks >= maxBlanks) break;
            } else {
              consecutiveBlanks = 0;
            }
            protectedSet.add(j);
            j++;
          }
          i = j;
          advanced = true;
        }
        break;
      }
    }

    if (!advanced) i++;
  }

  return protectedSet;
}


/**
 * Remove structural blocks identified by start/end marker patterns.
 * This enables removing multi-line sections (e.g. CodeRabbit review tables,
 * bot review sections) that individual line matching cannot catch.
 *
 * @param protectedLines - optional pre-computed set from `computeProtectedLines`.
 *   Lines in this set are immune from block removal.  When omitted, code fences
 *   are still protected via inline tracking (backward-compatible fallback).
 */
export function removeBlocks(lines: string[], ruleSet: CleanupRuleSet, protectedLines?: Set<number>): string[] {
  const patterns = ruleSet.blockPatterns;
  if (!patterns || patterns.length === 0) {
    return lines;
  }

  const result: string[] = [];
  let inCodeBlock = false;
  let i = 0;

  while (i < lines.length) {
    // Explicit protection: pre-computed protected lines are always kept and
    // can never trigger a block-removal start pattern.
    if (protectedLines?.has(i)) {
      result.push(lines[i]);
      i++;
      continue;
    }

    const trimmed = lines[i].trim();

    // Fallback code-block tracking (used when protectedLines is not passed)
    if (trimmed.startsWith('```')) {
      inCodeBlock = !inCodeBlock;
      result.push(lines[i]);
      i++;
      continue;
    }

    if (inCodeBlock) {
      result.push(lines[i]);
      i++;
      continue;
    }

    // Try to match a block start
    let matched = false;
    for (const bp of patterns) {
      if (bp.start.test(trimmed)) {
        const maxLen = bp.maxLines ?? 80;
        // Found a block start — scan forward for the end
        if (bp.end) {
          let j = i + 1;
          let found = false;
          while (j < lines.length && (j - i) < maxLen) {
            if (bp.end.test(lines[j].trim())) {
              found = true;
              j++; // skip the end line too
              break;
            }
            j++;
          }
          if (found) {
            i = j; // skip entire block
            matched = true;
          }
          // If end not found within maxLines, don't remove anything
        } else {
          // No end pattern — block extends to the first blank line, or to
          // `maxConsecutiveBlankLines` consecutive blank lines when > 1.
          const maxBlanks = bp.maxConsecutiveBlankLines ?? 1;
          let j = i + 1;
          let consecutiveBlanks = 0;
          while (j < lines.length && (j - i) < maxLen) {
            if (lines[j].trim() === '') {
              consecutiveBlanks++;
              if (consecutiveBlanks >= maxBlanks) {
                break;
              }
            } else {
              consecutiveBlanks = 0;
            }
            j++;
          }
          i = j; // skip block (the terminating blank line is kept on next iteration)
          matched = true;
        }
        break;
      }
    }

    if (!matched) {
      result.push(lines[i]);
      i++;
    }
  }

  return result;
}

/**
 * @param protectedLines - optional pre-computed set from `computeProtectedLines`.
 *   Lines in this set are always kept regardless of any removal rule.  When
 *   omitted, code fences are still protected via inline tracking (backward-
 *   compatible fallback).
 */
export function cleanMiddle(lines: string[], ruleSet: CleanupRuleSet, skipIntensive: boolean = false, protectedLines?: Set<number>): string[] {
  let inCodeBlock = false;

  return lines.filter((line, i) => {
    // Explicit protection: pre-computed protected lines are immune from all rules.
    if (protectedLines?.has(i)) return true;

    const trimmed = line.trim();
    
    if (trimmed.startsWith('```')) {
      inCodeBlock = !inCodeBlock;
      return true;
    }

    if (inCodeBlock) {
      return true;
    }

    if (trimmed === '') return true; // Keep blank lines for now, collapse later
    
    if (isPreserved(trimmed, ruleSet)) {
      return true;
    }

    if (ruleSet.removeAnywhereExactLines.includes(trimmed)) {
      return false;
    }

    if (!skipIntensive) {
      if (ruleSet.removeAnywhereContains.some(str => trimmed.includes(str))) {
        return false;
      }

      if (ruleSet.removeAnywhereRegexes?.some(regex => regex.test(trimmed))) {
        return false;
      }
    }

    return true;
  });
}

export function collapseBlankLines(lines: string[]): string[] {
  const result: string[] = [];
  let previousWasBlank = false;

  for (const line of lines) {
    const isBlank = line.trim() === '';
    if (isBlank) {
      if (!previousWasBlank) {
        result.push(line);
        previousWasBlank = true;
      }
    } else {
      result.push(line);
      previousWasBlank = false;
    }
  }

  // Trim leading/trailing blank lines
  while (result.length > 0 && result[0].trim() === '') {
    result.shift();
  }
  while (result.length > 0 && result[result.length - 1].trim() === '') {
    result.pop();
  }

  return result;
}

function formatSourceTypeLabel(type: SourceType): string {
  switch (type) {
    case 'github_pr':
      return 'GitHub Pull Request';
    case 'github_issue':
      return 'GitHub Issue';
    case 'docs':
      return 'Documentation';
    case 'article':
      return 'Article';
    case 'chat':
      return 'Chat';
    default:
      return 'Text';
  }
}

function normalizeMarkdownBody(lines: string[], type: SourceType): string[] {
  if (type === 'chat') {
    let inCodeBlock = false;
    return lines.map(line => {
      const trimmed = line.trim();
      if (!trimmed) {
        return '';
      }

      if (trimmed.startsWith('```')) {
        inCodeBlock = !inCodeBlock;
        return trimmed;
      }

      if (inCodeBlock) {
        return trimmed;
      }

      if (
        trimmed.startsWith('- ') ||
        trimmed.startsWith('* ') ||
        trimmed.startsWith('> ') ||
        trimmed.startsWith('#')
      ) {
        return trimmed;
      }

      return `- ${trimmed}`;
    });
  }

  return lines;
}

export function generateMarkdown(text: string, type: SourceType): string {
  const lines = normalizeText(text);
  if (lines.length === 0) {
    return '';
  }

  const title = `# Cleaned ${formatSourceTypeLabel(type)} Excerpt`;
  const body = normalizeMarkdownBody(lines, type).join('\n');

  return `${title}\n\n${body}`.trim();
}

export function generatePrompt(text: string, type: SourceType): string {
  const typeLabel = formatSourceTypeLabel(type);

  return [
    `Review the following cleaned ${typeLabel.toLowerCase()} excerpt.`,
    'Focus on the substantive content and ignore any residual site chrome.',
    '',
    text,
    '',
  ].join('\n');
}

export function cleanText(input: RawInput, ruleSetOverride?: CleanupRuleSet): CleanedResult {
  const isLargeText = input.rawText.length > 100000;
  const originalLines = normalizeText(input.rawText);
  
  let detectedType = input.sourceTypeHint;
  let preliminaryCleanedLines = originalLines;
  
  // If no hint is provided, we do a preliminary generic clean to help detection
  if (!detectedType) {
    preliminaryCleanedLines = trimPrefix(originalLines, GenericRuleSet);
    preliminaryCleanedLines = trimSuffix(preliminaryCleanedLines, GenericRuleSet);
    preliminaryCleanedLines = cleanMiddle(preliminaryCleanedLines, GenericRuleSet, isLargeText);
    preliminaryCleanedLines = collapseBlankLines(preliminaryCleanedLines);
    
    detectedType = detectSourceType(input.rawText, preliminaryCleanedLines.join('\n'));
  }

  const ruleSet = ruleSetOverride || getRuleSetForType(detectedType);
  
  let currentLines = originalLines;
  
  // If the ruleSet is GenericRuleSet, we already did the work!
  if (ruleSet === GenericRuleSet && !ruleSetOverride && !input.sourceTypeHint) {
    currentLines = preliminaryCleanedLines;
  } else {
    currentLines = trimPrefix(originalLines, ruleSet);
    currentLines = trimSuffix(currentLines, ruleSet);
    // Pipeline step 3: pre-compute protected lines BEFORE any removal step.
    // This guarantees that valuable context blocks (code fences, diff hunks,
    // rule-defined preserveBlockPatterns) can never be stripped by cleanup rules.
    const protectedLines = computeProtectedLines(currentLines, ruleSet);
    currentLines = removeBlocks(currentLines, ruleSet, protectedLines);
    currentLines = cleanMiddle(currentLines, ruleSet, isLargeText, protectedLines);
    currentLines = collapseBlankLines(currentLines);
  }

  const cleanedText = currentLines.join('\n');
  const removedLineCount = originalLines.length - currentLines.length;

  return {
    detectedType,
    cleanedText,
    markdownText: generateMarkdown(cleanedText, detectedType),
    llmPromptText: generatePrompt(cleanedText, detectedType),
    removedLineCount,
    warnings: removedLineCount === 0 && originalLines.length > 10 ? ['Low confidence: No lines were removed.'] : [],
  };
}

File: android/app/src/main/java/com/voku/textcleaner/core/Engine.kt
package com.voku.textcleaner.core

import com.voku.textcleaner.core.rules.ArticleRuleSet
import com.voku.textcleaner.core.rules.ChatRuleSet
import com.voku.textcleaner.core.rules.DocsRuleSet
import com.voku.textcleaner.core.rules.GenericRuleSet
import com.voku.textcleaner.core.rules.GitHubRuleSet

/**
 * Text cleanup engine.
 * Direct port of `src/core/engine.ts`.
 */
object Engine {

    fun getRuleSetForType(type: SourceType): CleanupRuleSet = when (type) {
        SourceType.GITHUB_PR, SourceType.GITHUB_ISSUE -> GitHubRuleSet
        SourceType.DOCS -> DocsRuleSet
        SourceType.ARTICLE -> ArticleRuleSet
        SourceType.CHAT -> ChatRuleSet
        SourceType.GENERIC -> GenericRuleSet
    }

    /**
     * Normalize line endings to `\n`, trim outer whitespace, split into lines.
     * Strips U+FFFC (Object Replacement Character) — web pages replace images/icons
     * with this character when copied as plain text; it is always noise.
     * Normalizes U+00A0 (No-Break Space) to a regular space — avoids mismatches
     * against rule strings that use ordinary spaces.
     */
    fun normalizeText(text: String): List<String> =
        text.replace("\r\n", "\n").trim().split("\n")
            .map { line -> line.replace("\uFFFC", "").replace("\u00A0", " ") }

    fun isPreserved(line: String, ruleSet: CleanupRuleSet): Boolean =
        ruleSet.preserveRegexes.any { it.containsMatchIn(line) }

    fun trimPrefix(lines: List<String>, ruleSet: CleanupRuleSet): List<String> {
        var unrecognizedCount = 0
        var lastJunkIndex = -1

        for (i in lines.indices) {
            val line = lines[i].trim()
            if (line.isEmpty()) continue

            var isJunk = false
            if (line in ruleSet.prefixExactLines) {
                isJunk = true
            } else if (ruleSet.prefixRegexes.any { it.containsMatchIn(line) }) {
                isJunk = true
            }

            if (isJunk) {
                lastJunkIndex = i
                unrecognizedCount = 0
            } else {
                if (isPreserved(line, ruleSet)) {
                    break
                }
                unrecognizedCount++
                if (unrecognizedCount > 5) {
                    break
                }
            }
        }

        return if (lastJunkIndex >= 0) {
            lines.subList(lastJunkIndex + 1, lines.size)
        } else {
            lines
        }
    }

    fun trimSuffix(lines: List<String>, ruleSet: CleanupRuleSet): List<String> {
        var unrecognizedCount = 0
        var firstJunkIndex = lines.size

        for (i in lines.indices.reversed()) {
            val line = lines[i].trim()
            if (line.isEmpty()) continue

            var isJunk = false
            if (line in ruleSet.suffixExactLines) {
                isJunk = true
            } else if (ruleSet.suffixRegexes.any { it.containsMatchIn(line) }) {
                isJunk = true
            }

            if (isJunk) {
                firstJunkIndex = i
                unrecognizedCount = 0
            } else {
                if (isPreserved(line, ruleSet)) {
                    break
                }
                unrecognizedCount++
                if (unrecognizedCount > 5) {
                    break
                }
            }
        }

        return if (firstJunkIndex < lines.size) {
            lines.subList(0, firstJunkIndex)
        } else {
            lines
        }
    }

    /**
     * Pre-compute a set of line indices that must survive all cleanup steps.
     *
     * This is the "protect specific blocks" stage that runs **before** any
     * removal step.  It identifies:
     *  1. Fenced code blocks (``` … ```) — always protected.
     *  2. Blocks matching [ruleSet.preserveBlockPatterns] — rule-defined content
     *     that gives the downstream LLM valuable context (e.g. diff hunks).
     *
     * Port of `computeProtectedLines()` in `src/core/engine.ts`.
     */
    fun computeProtectedLines(lines: List<String>, ruleSet: CleanupRuleSet): Set<Int> {
        val protectedSet = mutableSetOf<Int>()

        // Step 1 — always protect fenced code blocks
        var inCodeBlock = false
        for (i in lines.indices) {
            val trimmed = lines[i].trim()
            if (trimmed.startsWith("```")) {
                protectedSet.add(i) // fence line itself is always protected
                inCodeBlock = !inCodeBlock
            } else if (inCodeBlock) {
                protectedSet.add(i)
            }
        }

        // Step 2 — rule-defined preserve block patterns
        val patterns = ruleSet.preserveBlockPatterns
        if (patterns.isEmpty()) return protectedSet

        var i = 0
        while (i < lines.size) {
            if (i in protectedSet) { i++; continue }
            val trimmed = lines[i].trim()
            var advanced = false

            for (bp in patterns) {
                if (bp.start.containsMatchIn(trimmed)) {
                    val maxLen = bp.maxLines
                    protectedSet.add(i) // protect start line itself

                    if (bp.end != null) {
                        var j = i + 1
                        while (j < lines.size && (j - i) < maxLen) {
                            protectedSet.add(j)
                            if (bp.end.containsMatchIn(lines[j].trim())) {
                                i = j + 1
                                advanced = true
                                break
                            }
                            j++
                        }
                        if (!advanced) {
                            i = j // end not found — advance past
                            advanced = true
                        }
                    } else {
                        val maxBlanks = bp.maxConsecutiveBlankLines
                        var j = i + 1
                        var consecutiveBlanks = 0
                        while (j < lines.size && (j - i) < maxLen) {
                            if (lines[j].trim().isEmpty()) {
                                consecutiveBlanks++
                                if (consecutiveBlanks >= maxBlanks) break
                            } else {
                                consecutiveBlanks = 0
                            }
                            protectedSet.add(j)
                            j++
                        }
                        i = j
                        advanced = true
                    }
                    break
                }
            }

            if (!advanced) i++
        }

        return protectedSet
    }

    /**
     * Remove structural blocks identified by start/end marker patterns.
     * Enables removing multi-line sections (e.g. CodeRabbit review tables,
     * bot review sections) that individual line matching cannot catch.
     *
     * @param protectedLines optional pre-computed set from [computeProtectedLines].
     *   Lines in this set are immune from block removal.  When null, code fences
     *   are still protected via inline tracking (backward-compatible fallback).
     * Port of `removeBlocks()` in `src/core/engine.ts`.
     */
    fun removeBlocks(lines: List<String>, ruleSet: CleanupRuleSet, protectedLines: Set<Int>? = null): List<String> {
        val patterns = ruleSet.blockPatterns
        if (patterns.isEmpty()) return lines

        val result = mutableListOf<String>()
        var inCodeBlock = false
        var i = 0

        while (i < lines.size) {
            // Explicit protection: pre-computed protected lines are always kept
            // and can never trigger a block-removal start pattern.
            if (protectedLines != null && i in protectedLines) {
                result.add(lines[i])
                i++
                continue
            }

            val trimmed = lines[i].trim()

            // Fallback code-block tracking (used when protectedLines is not passed)
            if (trimmed.startsWith("```")) {
                inCodeBlock = !inCodeBlock
                result.add(lines[i])
                i++
                continue
            }

            if (inCodeBlock) {
                result.add(lines[i])
                i++
                continue
            }

            // Try to match a block start
            var matched = false
            for (bp in patterns) {
                if (bp.start.containsMatchIn(trimmed)) {
                    val maxLen = bp.maxLines
                    if (bp.end != null) {
                        // Scan forward for the end pattern
                        var j = i + 1
                        var found = false
                        while (j < lines.size && (j - i) < maxLen) {
                            if (bp.end.containsMatchIn(lines[j].trim())) {
                                found = true
                                j++ // skip the end line too
                                break
                            }
                            j++
                        }
                        if (found) {
                            i = j // skip entire block
                            matched = true
                        }
                        // If end not found within maxLines, don't remove anything
                    } else {
                        // No end pattern — block extends to the first blank line, or to
                        // maxConsecutiveBlankLines consecutive blank lines when > 1.
                        var j = i + 1
                        var consecutiveBlanks = 0
                        while (j < lines.size && (j - i) < maxLen) {
                            if (lines[j].trim().isEmpty()) {
                                consecutiveBlanks++
                                if (consecutiveBlanks >= bp.maxConsecutiveBlankLines) {
                                    break
                                }
                            } else {
                                consecutiveBlanks = 0
                            }
                            j++
                        }
                        i = j // skip block (terminating blank line kept on next iteration)
                        matched = true
                    }
                    break
                }
            }

            if (!matched) {
                result.add(lines[i])
                i++
            }
        }

        return result
    }

    /**
     * @param protectedLines optional pre-computed set from [computeProtectedLines].
     *   Lines in this set are always kept regardless of any removal rule.  When
     *   null, code fences are still protected via inline tracking (backward-
     *   compatible fallback).
     */
    fun cleanMiddle(
        lines: List<String>,
        ruleSet: CleanupRuleSet,
        skipIntensive: Boolean = false,
        protectedLines: Set<Int>? = null,
    ): List<String> {
        var inCodeBlock = false

        return lines.filterIndexed { i, line ->
            // Explicit protection: pre-computed protected lines are immune from all rules.
            if (protectedLines != null && i in protectedLines) return@filterIndexed true

            val trimmed = line.trim()

            if (trimmed.startsWith("```")) {
                inCodeBlock = !inCodeBlock
                return@filterIndexed true
            }

            if (inCodeBlock) {
                return@filterIndexed true
            }

            if (trimmed.isEmpty()) return@filterIndexed true // Keep blank lines for now, collapse later

            if (isPreserved(trimmed, ruleSet)) {
                return@filterIndexed true
            }

            if (trimmed in ruleSet.removeAnywhereExactLines) {
                return@filterIndexed false
            }

            if (!skipIntensive) {
                if (ruleSet.removeAnywhereContains.any { trimmed.contains(it) }) {
                    return@filterIndexed false
                }

                if (ruleSet.removeAnywhereRegexes.any { it.containsMatchIn(trimmed) }) {
                    return@filterIndexed false
                }
            }

            true
        }
    }

    fun collapseBlankLines(lines: List<String>): List<String> {
        val result = mutableListOf<String>()
        var previousWasBlank = false

        for (line in lines) {
            val isBlank = line.trim().isEmpty()
            if (isBlank) {
                if (!previousWasBlank) {
                    result.add(line)
                    previousWasBlank = true
                }
            } else {
                result.add(line)
                previousWasBlank = false
            }
        }

        // Trim leading blank lines
        while (result.isNotEmpty() && result[0].trim().isEmpty()) {
            result.removeAt(0)
        }
        // Trim trailing blank lines
        while (result.isNotEmpty() && result[result.size - 1].trim().isEmpty()) {
            result.removeAt(result.size - 1)
        }

        return result
    }

    private fun formatSourceTypeLabel(type: SourceType): String = when (type) {
        SourceType.GITHUB_PR -> "GitHub Pull Request"
        SourceType.GITHUB_ISSUE -> "GitHub Issue"
        SourceType.DOCS -> "Documentation"
        SourceType.ARTICLE -> "Article"
        SourceType.CHAT -> "Chat"
        SourceType.GENERIC -> "Text"
    }

    private fun normalizeMarkdownBody(lines: List<String>, type: SourceType): List<String> {
        if (type == SourceType.CHAT) {
            var inCodeBlock = false
            return lines.map { line ->
                val trimmed = line.trim()
                if (trimmed.isEmpty()) {
                    return@map ""
                }
                if (trimmed.startsWith("```")) {
                    inCodeBlock = !inCodeBlock
                    return@map trimmed
                }
                if (inCodeBlock) {
                    return@map trimmed
                }
                if (trimmed.startsWith("- ") ||
                    trimmed.startsWith("* ") ||
                    trimmed.startsWith("> ") ||
                    trimmed.startsWith("#")
                ) {
                    return@map trimmed
                }
                "- $trimmed"
            }
        }
        return lines
    }

    fun generateMarkdown(text: String, type: SourceType): String {
        val lines = normalizeText(text)
        if (lines.isEmpty()) {
            return ""
        }

        val title = "# Cleaned ${formatSourceTypeLabel(type)} Excerpt"
        val body = normalizeMarkdownBody(lines, type).joinToString("\n")

        return "$title\n\n$body".trim()
    }

    fun generatePrompt(text: String, type: SourceType): String {
        val preamble = generatePromptPreamble(type)

        return listOf(preamble, "", text, "").joinToString("\n")
    }

    private fun generatePromptPreamble(type: SourceType): String = when (type) {
        SourceType.GITHUB_PR ->
            "Review the following cleaned GitHub pull request excerpt. " +
            "This text is a copy-and-paste from a pull request; focus on the substantive content and ignore any residual site chrome. " +
            "Note that any code review feedback included in this content comes only from other LLMs and may simply be incorrect."
        SourceType.GITHUB_ISSUE ->
            "Review the following cleaned GitHub issue excerpt. " +
            "This text is a copy-and-paste from a GitHub issue; focus on the substantive content and ignore any residual site chrome."
        SourceType.DOCS ->
            "Review the following cleaned documentation excerpt. " +
            "This text is a copy-and-paste from documentation; focus on the substantive content and ignore any residual site chrome."
        SourceType.ARTICLE ->
            "Review the following cleaned article excerpt. " +
            "This text is a copy-and-paste from an article; focus on the substantive content and ignore any residual site chrome."
        SourceType.CHAT ->
            "Review the following cleaned chat excerpt. " +
            "This text is a copy-and-paste from a chat conversation; focus on the substantive content and ignore any residual site chrome."
        else ->
            "Review the following cleaned text excerpt. " +
            "Focus on the substantive content and ignore any residual site chrome."
    }

    fun cleanText(input: RawInput, ruleSetOverride: CleanupRuleSet? = null): CleanedResult {
        val isLargeText = input.rawText.length > 100_000
        val originalLines = normalizeText(input.rawText)

        var detectedType = input.sourceTypeHint
        var preliminaryCleanedLines = originalLines

        // If no hint is provided, do a preliminary generic clean to help detection
        if (detectedType == null) {
            preliminaryCleanedLines = trimPrefix(originalLines, GenericRuleSet)
            preliminaryCleanedLines = trimSuffix(preliminaryCleanedLines, GenericRuleSet)
            preliminaryCleanedLines = cleanMiddle(preliminaryCleanedLines, GenericRuleSet, isLargeText)
            preliminaryCleanedLines = collapseBlankLines(preliminaryCleanedLines)

            detectedType = Detector.detectSourceType(input.rawText, preliminaryCleanedLines.joinToString("\n"))
        }

        val ruleSet = ruleSetOverride ?: getRuleSetForType(detectedType)

        var currentLines: List<String>

        // If the ruleSet is GenericRuleSet, we already did the work!
        if (ruleSet == GenericRuleSet && ruleSetOverride == null && input.sourceTypeHint == null) {
            currentLines = preliminaryCleanedLines
        } else {
            currentLines = trimPrefix(originalLines, ruleSet)
            currentLines = trimSuffix(currentLines, ruleSet)
            // Pipeline step 3: pre-compute protected lines BEFORE any removal step.
            // Guarantees valuable context blocks (code fences, diff hunks,
            // rule-defined preserveBlockPatterns) can never be stripped by cleanup rules.
            val protectedLines = computeProtectedLines(currentLines, ruleSet)
            currentLines = removeBlocks(currentLines, ruleSet, protectedLines)
            currentLines = cleanMiddle(currentLines, ruleSet, isLargeText, protectedLines)
            currentLines = collapseBlankLines(currentLines)
        }

        val cleanedText = currentLines.joinToString("\n")
        val removedLineCount = originalLines.size - currentLines.size

        return CleanedResult(
            detectedType = detectedType,
            cleanedText = cleanedText,
            markdownText = generateMarkdown(cleanedText, detectedType),
            llmPromptText = generatePrompt(cleanedText, detectedType),
            removedLineCount = removedLineCount,
            warnings = if (removedLineCount == 0 && originalLines.size > 10) {
                listOf("Low confidence: No lines were removed.")
            } else {
                emptyList()
            },
        )
    }
}
